### DISCUSSION TITLE
How to properly call or link a StereoKitC function that utilizes `assets_execute_gpu`?

### DISCUSSION THREAD

**Discussion by austinbhale** [StereoKit/StereoKit#624] [Q&A]
## Context - HL2, ARM
I have a C API that statically links to the StereoKitC API via CMake, so I've been able to use the SK libraries with no issue. However, I've been having trouble with `sk::assets_execute_gpu`. I also import both StereoKit and my C DLLs in C#.

## Example in C
```
mesh_t mesh;
void test_mesh_set() {
    mesh = mesh_create();
    vind_t inds[4] = { 1, 2, 3, 4 };
    mesh_set_inds(mesh, inds, 4);
}
```
---

I called this test method anywhere in C#, whether it be after SK's first step, on the main thread, background thread, etc. and get an access violation when locking the mutex and it shows a Win32 CriticalSelection error from tinycthread:
https://github.com/StereoKit/StereoKit/blob/aae289a27ad8275466da31e74a66ecf9814889a5/StereoKitC/asset_types/assets.cpp#L450

What's possibly interesting is that the stack trace is calling these SK functions through my DLL that statically links SK rather than the SK DLL?

![image](https://user-images.githubusercontent.com/32556981/234821100-1a376983-4edf-4a00-9944-b926cf2017a1.png)
![image](https://user-images.githubusercontent.com/32556981/234822048-05f35ce8-aabe-4589-a448-1ee7fe82b585.png)

Could this be some race for critical section access? Or is it not recommended to statically link SK to a custom shared DLL and use SK's shared DLL simultaneously?

**Comment by austinbhale** [comment]
I read that Windows doesn't let much global var sharing between DLLs and critical sections are local to a single process. So in order to access the mutex state, I imported the StereoKitC DLL and got the previous example working:

![image](https://user-images.githubusercontent.com/32556981/234837557-cab04139-4c12-4bb7-8bbd-c2641d223f56.png)

I guess a DLL on top of a static link to SK is only good for stateless usage. I'll just use the shared library natively as well.

Is there already an established way to import the SK functions similar to the [loading of open gl functions](https://github.com/StereoKit/StereoKit/blob/e72bf7d29db8b1ef79092cdf7dd8e8b851e669b5/StereoKitC/libraries/sk_gpu.h#L2522)?

**Comment by austinbhale** [comment]
In order to automatically find the function addresses in StereoKitC, one of the simplest solutions is just enabling shared libraries: `set(SK_BUILD_SHARED_LIBS ON CACHE INTERNAL "")` to reuse the same DLL

The main reason I wanted to link as a static library is so I could have access to the helpful sk_memory methods. I could always link it statically and then import every function manually like above but that's a bit tedious and a larger file. Or would it be possible to mark these public headers with `SK_API` as well? For example, `SK_API void *sk_malloc (size_t bytes);`?

**Comment by maluoi** [comment]
To be honest, I've never gone through this sort of flow before, but mixing static and dynamic linking like this does sound like it could be quite a minefield! In particular, if you're loading StereoKitC.dll _and_ a library that static links to a StereoKitC.lib, I guess I'd probably expect that project to have duplicate copies of StereoKit code in it?

I've also never even thought to set up StereoKit nicely for dynamically loading function pointers like with GetProcAddress, so I don't have anything existing to point you to there either! It really would just be a lot of typing. I might think about that for some of the auto-gen stuff later though.

I'm also kinda interested, why the sk_memory functions? These are basically just light wrappers around malloc.h functions! The only reason I really have them is for suppressing warnings, and debugging purposes with `SK_DEBUG_MEM`. The core functionality should be pretty trivially replicable!

**Reply by austinbhale** [reply]
> duplicate copies of StereoKit code in it?

Yep that's exactly what was going on. The C# app referenced the StereoKitC DLL and my DLL linked StereoKitC statically 😶‍🌫️ but you kind of need the DLL for C# so the static link doesn't make much sense on top of that. However, linking it as a shared lib with my C project works really well

> dynamically loading function pointers like with GetProcAddress

So much typing lol probably not a fun one to add. It's nice to know dynamically loading SK functions is an option though for other potential libs

> why the sk_memory functions?

Just thought as these as an example for futureproofing what can be borrowed from SK. Is there a grouping for what is deemed as part of the C API like [this grouping `SK_SRC_CODE`](https://github.com/StereoKit/StereoKit/blob/e72bf7d29db8b1ef79092cdf7dd8e8b851e669b5/CMakeLists.txt#LL174C5-L174C16)? No worries if not, I don't mind replicating those portions

**Reply by maluoi** [reply]
The only files I use for the public API are `stereokit.h` and `stereokit_ui.h`! Any other file I would consider to be "peeking under the covers" and are subject to potentially massive change with no changelog notes. So I'd generally be pretty cautious about using anything else directly :)

**Reply by austinbhale** [reply]
Perfect, thanks so much! 😄
